File: planik-android/app/src/main/java/pl/planik/domain/service/UserService.kt
package pl.planik.domain.service

import pl.planik.domain.model.NewUser
import pl.planik.domain.model.User
import pl.planik.domain.source.UserLocalSource
import javax.inject.Inject

class UserService @Inject constructor(
  private val userLocalSource: UserLocalSource
) {

  fun hasUser(): Boolean = userLocalSource.hasUser()

  suspend fun getCurrentUser(): User? = userLocalSource.getUser()

  suspend fun createUser(nickname: String): User? {
    return userLocalSource.createUser(newUser = NewUser(nickname = nickname))
  }
}

File: planik-flutter/planik/lib/domain/service/user_service.dart
import 'package:fpdart/fpdart.dart';
import 'package:planik/domain/failure/general_failure.dart';
import 'package:planik/domain/model/user.dart';
import 'package:planik/domain/source/user_local_source.dart';
import 'package:uuid/uuid.dart';

class UserService {
  const UserService({
    required UserLocalSource userLocalSource,
  }) : _userLocalSource = userLocalSource;

  final UserLocalSource _userLocalSource;

  TaskEither<GeneralFailure, bool> hasUser() => _userLocalSource.hasUser();

  TaskEither<GeneralFailure, bool> hasPlan() => _userLocalSource.hasPlan();

  TaskEither<GeneralFailure, User> getCurrentUser() => _userLocalSource.getCurrentUser();

  TaskEither<GeneralFailure, User> createUser({required String nickname}) {
    return _userLocalSource.create(
      nickname: nickname,
      uid: Uuid(
        options: {'seedBytes': nickname.codeUnits},
      ).v4(),
    );
  }
}
